crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2026,6 +2030,109 @@ impl<'db> Class<'db> {
+    /// Only call this function from queries in the same file or your
+    /// query depends on the AST of another file (bad!).
+    fn explicit_metaclass(self, db: &'db dyn Db) -> Option<Type<'db>> {
+        let class_stmt = self.node(db);
+        let metaclass = class_stmt
+            .keywords()
+            .iter()
+            .find(|keyword| keyword.arg.as_ref().is_some_and(|arg| arg == "metaclass"))
+            .map(|keyword| &keyword.value);
+        if class_stmt.type_params.is_some() {
+            // when we have a specialized scope, we'll look up the inference
+            // within that scope
+            let model = SemanticModel::new(db, self.file(db));
+            metaclass.map(|base| base.ty(&model))
+        } else {
+            // Otherwise, we can do the lookup based on the definition scope
+            let class_definition = semantic_index(db, self.file(db)).definition(class_stmt);
+            metaclass.map(|base_node| definition_expression_ty(db, class_definition, base_node))
+        }
+    }
+
+    /// Return the metaclass of this class.
+    #[salsa::tracked]
+    pub(crate) fn metaclass(self, db: &'db dyn Db) -> Result<Type<'db>, MetaclassError<'db>> {
+        /// Infer the metaclass of a class, tracking the classes that have been visited to detect
+        /// cyclic definitions.
+        fn infer<'db>(
+            db: &'db dyn Db,
+            class: Class<'db>,
+            watcher: &mut Watcher<Class<'db>>,
+        ) -> Result<Type<'db>, MetaclassError<'db>> {
+            let mut base_classes = class
+                .explicit_bases(db)
+                .iter()
+                .copied()
+                .filter_map(Type::into_class_literal);
+
+            // Identify the class's own metaclass (or take the first base class's metaclass).
+            let metaclass = if let Some(metaclass) = class.explicit_metaclass(db) {
+                metaclass
+            } else if let Some(base_class) = base_classes.next() {
+                // Each base must be considered in isolation.
+                let watcher_len = watcher.len();
+                if !watcher.insert(base_class.class) {
+                    return Err(MetaclassError {
+                        kind: MetaclassErrorKind::CyclicDefinition,
+                    });
+                }
+                let metaclass = infer(db, base_class.class, watcher)?;
+                watcher.truncate(watcher_len);
+                metaclass
+            } else {
+                KnownClass::Type.to_class(db)
+            };
+
+            let Type::ClassLiteral(mut candidate) = metaclass else {
+                // If the metaclass is not a class, return it directly.
+                return Ok(metaclass);
+            };
+
+            // Reconcile all base classes' metaclasses with the candidate metaclass.
+            //
+            // See: https://github.com/python/cpython/blob/83ba8c2bba834c0b92de669cac16fcda17485e0e/Objects/typeobject.c#L3629-L3663
```

Comment:
Is there a reference in the language spec we could also link to somewhere? Maybe in the data model?